ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -584,36 +628,22 @@ void writeIndexSegmentToFile(Offset safeEndPoint) throws IOException, StoreExcep
           writer.writeShort(resetKey.getSecond().ordinal());
         }
 
-        // NOTE: In the event of a crash, it is possible that there is a part of the log that is not covered by the
-        // index. This happens due to the fact that a DELETE that occurs in the same segment as a PUT overwrites the
-        // PUT entry. Consider the following case:-
-        // (entries are of the form ID:TYPE:START_OFFSET-END_OFFSET)
-        // This is the order of operations
-        // A:PUT:0-100
-        // B:PUT:101-200
-        // A:DELETE:201-250
-        // These are the entries in the index segment
-        // A:DELETE:201-250
-        // B:PUT:101-200
-        // If safeEndPoint < 250, then B:PUT:101-200 will be written but not A:DELETE:201-250. If the process were to
-        // crash at this point, the index end offset would be 200 and the span 0-100 would not be represented in the
-        // index.
-        // write the entries
         byte[] maxPaddingBytes = null;
         if (getVersion() == PersistentIndex.VERSION_2) {
           maxPaddingBytes = new byte[persistedEntrySize - valueSize];
         }
-        for (Map.Entry<StoreKey, IndexValue> entry : index.entrySet()) {
-          if (entry.getValue().getOffset().getOffset() + entry.getValue().getSize() <= safeEndPoint.getOffset()) {
-            writer.write(entry.getKey().toBytes());
-            writer.write(entry.getValue().getBytes().array());
-            if (getVersion() == PersistentIndex.VERSION_2) {
-              // Add padding if necessary
-              writer.write(maxPaddingBytes, 0, persistedEntrySize - (entry.getKey().sizeInBytes() + valueSize));
+        for (Map.Entry<StoreKey, NavigableSet<IndexValue>> entry : index.entrySet()) {
```

Comment:
there should be a clever way of using java 8 streams here to flatten the map, which can perhaps simplify this code. For example, `index.entrySet().stream().flatMap(e -> e.getValue().stream())` flattens and creates a stream of all individual entries. @cgtz any ideas?

Reply:
I'll try to see if anything can be done

Reply:
the `Consumer` in `forEach` can't throw checked exceptions (the `writer.write()` throws `IOException`) so the final code (if we want to wrap the `IOException` in a `IllegalStateException`) is not much simpler or concise than what is there now (might even look more complicated and be more lines overall). I don't flattening improves it much either. Let me know if you or @cgtz have other ideas I can try.